File: model_1/CS3233_2025_Midterm/imperfection_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

const int MAX_V = 300000;
const int MAX_C = 300000;

class Fenw {
public:
    int n;
    vector<long long> tree;

    Fenw(int size) : n(size) {
        tree.assign(n+1, 0);
    }

    void update(int index, long long delta) {
        while (index <= n) {
            tree[index] += delta;
            index += index & -index;
        }
    }

    long long query(int index) {
        if (index <= 0) return 0;
        if (index > n) index = n;
        long long s = 0;
        while (index) {
            s += tree[index];
            index -= index & -index;
        }
        return s;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> A(n), B(n);
    for (int i = 0; i < n; i++) cin >> A[i];
    for (int i = 0; i < n; i++) cin >> B[i];

    vector<int> F(MAX_V+2, 0);
    vector<long long> sum_cost(MAX_V+2, 0);
    vector<vector<int>> costs_by_value(MAX_V+2);

    for (int i = 0; i < n; i++) {
        if (A[i] > MAX_V) continue;
        F[A[i]]++;
        sum_cost[A[i]] += B[i];
        costs_by_value[A[i]].push_back(B[i]);
    }

    vector<long long> suffix_sum(MAX_V+3, 0);
    for (int i = MAX_V; i >= 1; i--) {
        suffix_sum[i] = suffix_sum[i+1] + sum_cost[i];
    }

    vector<int> total_count_below(MAX_V+2, 0);
    for (int x = 1; x <= MAX_V; x++) {
        total_count_below[x] = total_count_below[x-1] + F[x-1];
    }

    vector<vector<long long>> prefix_x_arr(MAX_V+2);
    for (int v = 1; v <= MAX_V; v++) {
        if (F[v] > 0) {
            sort(costs_by_value[v].begin(), costs_by_value[v].end());
            vector<long long> arr(F[v]+1);
            arr[0] = 0;
            for (int i = 1; i <= F[v]; i++) {
                arr[i] = arr[i-1] + costs_by_value[v][i-1];
            }
            prefix_x_arr[v] = arr;
        }
    }

    Fenw cnt_tree(MAX_C);
    Fenw sum_tree(MAX_C);

    long long candidate_ans = LLONG_MAX;

    for (int x = 1; x <= MAX_V; x++) {
        if (x >= 2) {
            for (int c : costs_by_value[x-1]) {
                cnt_tree.update(c, 1);
                sum_tree.update(c, c);
            }
        }

        if (F[x] == 0) continue;

        long long fixed_cost = suffix_sum[x+1];
        int k0 = F[x];
        int t0 = total_count_below[x];
        int D0 = max(0, t0 - k0 + 1);

        if (D0 == 0) {
            if (fixed_cost < candidate_ans) {
                candidate_ans = fixed_cost;
            }
        } else {
            long long best_step2 = LLONG_MAX;
            for (int b = 0; b < k0; b++) {
                long long a = (long long)D0 + b;

                int low = 1, high = MAX_C;
                int T = MAX_C;
                while (low <= high) {
                    int mid = (low + high) >> 1;
                    long long cnt_here = cnt_tree.query(mid);
                    if (cnt_here >= a) {
                        T = mid;
                        high = mid - 1;
                    } else {
                        low = mid + 1;
                    }
                }

                long long cnt1 = cnt_tree.query(T-1);
                long long sum1 = sum_tree.query(T-1);
                long long s_nonx = sum1 + (a - cnt1) * T;

                long long s_x = prefix_x_arr[x][b];
                long long total_var = s_nonx + s_x;
                if (total_var < best_step2) {
                    best_step2 = total_var;
                }
            }
            long long total_cost = fixed_cost + best_step2;
            if (total_cost < candidate_ans) {
                candidate_ans = total_cost;
            }
        }
    }

    cout << candidate_ans << endl;

    return 0;
}

File: model_1/CS3233_2025_Midterm/imperfection_1_solution.py
import sys

class Fenw:
    def __init__(self, size):
        self.n = size
        self.tree = [0] * (self.n + 1)
    
    def update(self, index, delta):
        while index <= self.n:
            self.tree[index] += delta
            index += index & -index
            
    def query(self, index):
        if index <= 0:
            return 0
        if index > self.n:
            index = self.n
        s = 0
        while index:
            s += self.tree[index]
            index -= index & -index
        return s

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    
    n = int(data[0])
    A = list(map(int, data[1:1+n]))
    B = list(map(int, data[1+n:1+2*n]))
    
    MAX_V = 300000
    MAX_C = 300000
    
    F = [0] * (MAX_V + 2)
    sum_cost = [0] * (MAX_V + 2)
    costs_by_value = [[] for _ in range(MAX_V + 2)]
    
    for i in range(n):
        a_val = A[i]
        b_val = B[i]
        if a_val > MAX_V:
            continue
        F[a_val] += 1
        sum_cost[a_val] += b_val
        costs_by_value[a_val].append(b_val)
    
    suffix_sum = [0] * (MAX_V + 3)
    for i in range(MAX_V, 0, -1):
        suffix_sum[i] = suffix_sum[i+1] + sum_cost[i]
    
    total_count_below = [0] * (MAX_V + 2)
    for x in range(1, MAX_V + 1):
        total_count_below[x] = total_count_below[x-1] + F[x-1]
    
    prefix_x_arr = [None] * (MAX_V + 2)
    for v in range(1, MAX_V + 1):
        if F[v] > 0:
            lst = sorted(costs_by_value[v])
            arr = [0] * (F[v] + 1)
            for i in range(1, F[v] + 1):
                arr[i] = arr[i-1] + lst[i-1]
            prefix_x_arr[v] = arr
    
    cnt_tree = Fenw(MAX_C)
    sum_tree = Fenw(MAX_C)
    candidate_ans = 10**18
    
    for x in range(1, MAX_V + 1):
        if x >= 2:
            for c in costs_by_value[x-1]:
                cnt_tree.update(c, 1)
                sum_tree.update(c, c)
                
        if F[x] == 0:
            continue
            
        fixed_cost = suffix_sum[x+1]
        k0 = F[x]
        t0 = total_count_below[x]
        
        if k0 > t0:
            total_cost = fixed_cost
            if total_cost < candidate_ans:
                candidate_ans = total_cost
        else:
            best_var = 10**18
            for b in range(0, k0):
                a_min = t0 - k0 + b + 1
                if a_min > t0:
                    break
                low, high = 1, MAX_C
                while low < high:
                    mid = (low + high) // 2
                    if cnt_tree.query(mid) >= a_min:
                        high = mid
                    else:
                        low = mid + 1
                T = low
                cnt1 = cnt_tree.query(T-1)
                sum1 = sum_tree.query(T-1)
                nonx_cost = sum1 + T * (a_min - cnt1)
                x_cost = prefix_x_arr[x][b]
                total_var = nonx_cost + x_cost
                if total_var < best_var:
                    best_var = total_var
                    
            total_cost = fixed_cost + best_var
            if total_cost < candidate_ans:
                candidate_ans = total_cost
                
    print(candidate_ans)

if __name__ == '__main__':
    main()